### PR TITLE
TAJO-1735: Implement MetadataProvider and LinkedMetadataManager.

### PR DESCRIPTION
**Pull request by hyunsik**
[Review on Reviewable](https://reviewable.io/reviews/apache/tajo/673)

### CODE REVIEW

**Comment by blrunner** [comment]
Hi @hyunsik 

Could you rebase it? :)

**Comment by blrunner** [comment]
Overall, it looks good to me. And if you add more unit test cases for InsufficientPrivilegeException, it would be more nice.

File: tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java [hunk 1]
```diff
@@ -97,19 +103,17 @@
   public CatalogServer() throws IOException {
     super(CatalogServer.class.getName());
     this.handler = new CatalogProtocolHandler();
+    this.linkedMetadataManager = new LinkedMetadataManager(Collections.EMPTY_LIST);
     this.builtingFuncs = new ArrayList<FunctionDesc>();
   }
 
-  public CatalogServer(Collection<FunctionDesc> sqlFuncs) throws IOException {
-    this();
+  public CatalogServer(Set<MetadataProvider> metadataProviders, Collection<FunctionDesc> sqlFuncs) throws IOException {
```
Comment (blrunner) [inline tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java]: Why don't you implement a constructor which needs only MetadataProvider?

File: tajo-catalog/tajo-catalog-common/src/main/java/org/apache/tajo/catalog/CatalogService.java [hunk 2]
```diff
@@ -24,6 +24,7 @@
 import org.apache.tajo.catalog.proto.CatalogProtos.*;
 import org.apache.tajo.common.TajoDataTypes.DataType;
 
+import javax.annotation.Nullable;
```
Comment (blrunner) [inline tajo-catalog/tajo-catalog-common/src/main/java/org/apache/tajo/catalog/CatalogService.java]: It looks like unused package.

File: tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java [hunk 3]
```diff
@@ -304,9 +295,25 @@ public GetTablespaceListResponse getAllTablespaces(RpcController controller, Nul
         throws ServiceException {
       rlock.lock();
       try {
+
+        // retrieves tablespaces from catalog store
+        final List<TablespaceProto> tableSpaces = Lists.newArrayList(store.getTablespaces());
+
+        // retrieves tablespaces from linked meta data
+        tableSpaces.addAll(Collections2.transform(linkedMetadataManager.getTablespaces(),
```
Comment (blrunner) [inline tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java]: Could you explain why you use Collections2?

File: tajo-catalog/tajo-catalog-server/src/test/java/org/apache/tajo/catalog/TestLinkedMetadataManager.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    assertEquals("http://space1/x", space1.getUri());
+
+    CatalogProtos.TablespaceProto space2 = catalog.getTablespace("space2");
+    assertEquals("space2", space2.getSpaceName());
+    assertEquals("http://space2/y", space2.getUri());
+  }
+
+  @Test
+  public void testGetTablespaces() throws Exception {
+    Collection<String> names = Collections2.transform(catalog.getAllTablespaces(), new Function<CatalogProtos.TablespaceProto, String>() {
```
Comment (blrunner) [inline tajo-catalog/tajo-catalog-server/src/test/java/org/apache/tajo/catalog/TestLinkedMetadataManager.java]: Exceeded character limit for one line.

Comment (hyunsik) [inline tajo-catalog/tajo-catalog-server/src/test/java/org/apache/tajo/catalog/TestLinkedMetadataManager.java]: fixed

File: tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java (hunk 3 above)
Comment (hyunsik) [inline tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java]: Some functional methods like transform and filter are in guava Collections2 class.

File: tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java (hunk 1 above)
Comment (hyunsik) [inline tajo-catalog/tajo-catalog-server/src/main/java/org/apache/tajo/catalog/CatalogServer.java]: It is because the constructor is unnecessary.

**Comment by hyunsik** [comment]
rebased and reflected the comments.

**Comment by blrunner** [comment]
+1
LGTM.:)

**Pull request closed without merging**